features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularStack.cpp

```diff
@@ -0,0 +1,282 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "UBLOX_N2XX_CellularStack.h"
+#include "CellularUtil.h"
+#include <stdlib.h>
+
+using namespace mbed;
+using namespace mbed_cellular_util;
+
+UBLOX_N2XX_CellularStack::UBLOX_N2XX_CellularStack(ATHandler &atHandler, int cid, nsapi_ip_stack_t stack_type) : AT_CellularStack(atHandler, cid, stack_type)
+{
+    // URC handlers for sockets
+    _at.set_urc_handler("+NSONMI:", callback(this, &UBLOX_N2XX_CellularStack::NSONMI_URC));
+}
+
+UBLOX_N2XX_CellularStack::~UBLOX_N2XX_CellularStack()
+{
+}
+
+nsapi_error_t UBLOX_N2XX_CellularStack::socket_listen(nsapi_socket_t handle, int backlog)
+{
+    return NSAPI_ERROR_UNSUPPORTED;
+}
+
+nsapi_error_t UBLOX_N2XX_CellularStack::socket_accept(void *server, void **socket, SocketAddress *addr)
+{
+    return NSAPI_ERROR_UNSUPPORTED;
+}
+
+// Callback for Socket Read URC.
+void UBLOX_N2XX_CellularStack::NSONMI_URC()
+{
+    int a, b;
+    CellularSocket *socket;
+
+    a = _at.read_int();
+    b = _at.read_int();
+
+    socket = find_socket(a);
+    if (socket != NULL) {
+        socket->pending_bytes = b;
+        // No debug prints here as they can affect timing
+        // and cause data loss in UARTSerial
+        if (socket->_cb != NULL) {
+            socket->_cb(socket->_data);
+        }
+    }
+}
+
+int UBLOX_N2XX_CellularStack::get_max_socket_count()
+{
+    return UBLOX_MAX_SOCKET;
+}
+
+bool UBLOX_N2XX_CellularStack::is_protocol_supported(nsapi_protocol_t protocol)
+{
+    return (protocol == NSAPI_UDP);
+}
+
+nsapi_error_t UBLOX_N2XX_CellularStack::create_socket_impl(CellularSocket *socket)
+{
+    int sock_id = 0;
+    int localport = socket->localAddress.get_port();
+
+    if(localport == 5683 && localport < 0 && localport > 65535) {
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+
+    _at.lock();
+    _at.cmd_start("AT+NSOCR=\"DGRAM\",17,");
+    _at.write_int(localport);
+    _at.write_int(1);
+    _at.cmd_stop();
+
+    _at.resp_start();
+    sock_id = _at.read_int();
+    _at.resp_stop();
+
+    if ((_at.get_last_error() != NSAPI_ERROR_OK) || (sock_id == -1)) {
+        _at.unlock();
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+    _at.unlock();
+
+    // Check for duplicate socket id delivered by modem
+    for (int i = 0; i < UBLOX_MAX_SOCKET; i++) {
+        CellularSocket *sock = _socket[i];
+        if (sock && sock->created && sock->id == sock_id) {
+            return NSAPI_ERROR_NO_SOCKET;
+        }
+    }
+
+    socket->id = sock_id;
+    socket->created = true;
+
+    return NSAPI_ERROR_OK;
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_sendto_impl(CellularSocket *socket, const SocketAddress &address,
+        const void *data, nsapi_size_t size)
+{
+    int sent_len = 0, len = 0;
+    char *dataStr = (char *) malloc((size * 2) + 1);
+    memset(dataStr, 0, size*2+1);
+
+    char_str_to_hex_str((const char*)data, size, dataStr);
```

Comment:
I can't find the char_str_to_hex_str() and the reverse functions in here: am I missing something?

Reply:
These functions are present under namespace mbed_cellular_util. 

https://github.com/u-blox/mbed-os/blob/master/features/cellular/framework/common/CellularUtil.cpp

Reply:
It's in here:
mbed-os/features/cellular/framework/common/CellularUtil.cpp